src/core/app.lua

```diff
@@ -124,6 +128,88 @@ function configure (new_config)
    counter.add(configs)
 end
 
+-- Claims a name for a program so it can be identified by name by other processes.
+--
+-- The name given to the function must be unique, if a name has been used before by
+-- an active process the function will error displaying an appropriate error message.
+-- Any other process can enumerate programs by their name and PID using enumerate_named_programs.
+function claim_name(name)
+   function name_from_path(dir)
+      -- Extracts last file/folder name from path
+      -- e.g. "/var/run/snabb/hello" -> "hello"
+      local ds, de = string.find(dir, "/[^/]*$")
+      return string.sub(dir, ds+1, de)
+   end
+
+   configuration[name] = name
+   local namedir = "by-name/" .. name
+   local namedirqualified = named_program_root .. "/" .. name
+   local piddir = shm.root .. "/" .. S.getpid()
+   local backlinkdir = piddir.."/name-backref"
+
+   -- Verify that the by-name directory exists.
+   shm.mkdir(namedir)
+
+   -- Verify that we've not already claimed a name
+   if S.lstat(backlinkdir) then
+      -- Read name of program for error message.
+      local fnamedir = S.readlink(backlinkdir)
+      local fname = name_from_path(fnamedir)
+      error("Program already claimed name '"..fname.."', cannot claim name '"..name.."'")
+   end
+
+   -- Check if a symlink to a named program with the same name exists, if not, unlink it.
```

Comment:
Let's elide this section (lines 161-173).  Even with this section there's a TOCTTOU bug, but one which is robustly handled by the S.symlink call later, so this check is needless.  In theory we never get into this situation anyway because of the supervisor process.